db/batch/batch_impl.go

```diff
@@ -96,19 +96,30 @@ func (b *baseKVStoreBatch) Entry(index int) (*WriteInfo, error) {
 func (b *baseKVStoreBatch) SerializeQueue(serialize WriteInfoSerialize, filter WriteInfoFilter) []byte {
 	b.mutex.Lock()
 	defer b.mutex.Unlock()
-	// 1. This could be improved by being processed in parallel
-	// 2. Digest could be replaced by merkle root if we need proof
-	bytes := make([]byte, 0)
+	// 1. Digest could be replaced by merkle root if we need proof
+	bytesChan := make(chan []byte, len(b.writeQueue))
+	wg := sync.WaitGroup{}
+	wg.Add(len(b.writeQueue))
 	for _, wi := range b.writeQueue {
-		if filter != nil && filter(wi) {
-			continue
-		}
-		if serialize != nil {
-			bytes = append(bytes, serialize(wi)...)
-		} else {
-			bytes = append(bytes, wi.Serialize()...)
-		}
+		go func(info *WriteInfo) {
+			defer wg.Done()
+			if filter != nil && filter(info) {
+				return
+			}
+			if serialize != nil {
+				bytesChan <- serialize(info)
+			} else {
+				bytesChan <- info.Serialize()
+			}
+		}(wi)
 	}
+	wg.Wait()
+
+	bytes := make([]byte, 0)
+	for itemBytes := range bytesChan {
+		bytes = append(bytes, itemBytes...)
+	}
```

Comment:
in general this is fine, but probably won't work for our case. Because the total bytes will be hashed to serve as a digest of the batch's content, which is used later in validation. So the bytes has to be in fixed order, and working parallel breaks this order

Reply:
I see, does this mean the PR is no longer needed? Thanks

Reply:
@envestcc modified it so the writing of the bytes matches the order 👍

Reply:
does the `Unlock` can be called earlier? It seems like that it can be placed just before `wg.Wait()`.

but it seems that this is unrelated to the current issue. It may not improve the serialization process. Perhaps it can be discussed in a new issue.